gwctl/pkg/resourcediscovery/discoverer.go

```diff
@@ -284,19 +303,56 @@ func (d Discoverer) discoverHTTPRoutesFromBackends(ctx context.Context, resource
 	}
 
 	for _, httpRoute := range httpRoutes {
-		var found bool
+		// An HTTPRoute will be included in the resourceModel if it references some
+		// Backend which already exists in the resourceModel.
+		var includeRouteInResourceModel bool
+
 		for _, backendRef := range relations.FindBackendRefsForHTTPRoute(httpRoute) {
+			// Check if the referenced backend exists in the resourceModel.
 			backendID := BackendID(backendRef.Group, backendRef.Kind, backendRef.Namespace, backendRef.Name)
-			_, ok := resourceModel.Backends[backendID]
+			backendNode, ok := resourceModel.Backends[backendID]
 			if !ok {
 				continue
 			}
-			found = true
+
+			// Ensure that if this is a cross namespace reference, then it is accepted
+			// through some ReferenceGrant.
+			if httpRoute.GetNamespace() != backendRef.Namespace {
+				httpRouteRef := common.ObjRef{
+					Group:     httpRoute.GroupVersionKind().Group,
+					Kind:      httpRoute.GroupVersionKind().Kind,
+					Name:      httpRoute.GetName(),
+					Namespace: httpRoute.GetNamespace(),
+				}
+				var referenceAccepted bool
+				for _, referenceGrantNode := range backendNode.ReferenceGrants {
+					if relations.ReferenceGrantAccepts(*referenceGrantNode.ReferenceGrant, httpRouteRef) {
+						referenceAccepted = true
+						break
+					}
+				}
+				if !referenceAccepted {
+					err := ReferenceNotPermittedError{ReferenceFromTo: ReferenceFromTo{
+						ReferringObject: common.ObjRef{Kind: "HTTPRoute", Name: httpRoute.GetName(), Namespace: httpRoute.GetNamespace()},
+						ReferredObject:  backendRef,
+					}}
+					backendNode.Errors = append(backendNode.Errors, err)
+					klog.V(1).Info(err)
```

Comment:
Wouldn't this output immediately? It seems like it might be more ideal to have an overall output that lists warnings like this at the bottom of the entire output instead of interspersed throughout.

Reply:
Ahh I think the key piece is the verbosityLevel=1 (which won't get printed by default).

You are absolutely right this should not get interpersed with all the other output -- it will only happen people set that verbosity level `gwctl get gateways -v=1`. This is less user facing and more for enabling us to write better tests (as I've enabled this by default in tests and we should see what decisions the code takes along the way)

On the other part, of listing warnings together -- yes that's absolutely the plan. The new "Errors" introduced in this PR on the resource are just for that purpose i.e. when some does something like `gwctl describe gateways`, they would see an `Errors` or `Analysis` section about such things (it's available in the data-model and simply needs to be exposed on the Describe view now)